### PR TITLE
first cut at redis support

### PR DESCRIPTION
**Pull request by memory**
- add a redis store implementing stores.Storage
- add config file support to pick a storage backend
- add config file support to set redis host:port and password

Also:

- add docker_releases to .gitignore

### CODE REVIEW

File: stores/redis/redis.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/mxschmitt/golang-url-shortener/stores/shared"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	entryPathPrefix     = string("entry:")       // prefix for path-to-url mappings
```
Comment (mxschmitt) [inline stores/redis/redis.go]: Why always `string()`?
As far as I know `"foo"` is per default a string

File: stores/redis/redis.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+		return nil, errors.Wrap(err, "Could not connect to redis db0")
+	}
+	ret := &RedisStore{c: c}
+	return ret, nil
+}
+
+func (r *RedisStore) keyExists(key string) (exists bool, err error) {
+	var result *redis.IntCmd
+	logrus.Debugf("Checking for existence of key: %s", key)
+	result = r.c.Exists(key)
```
Comment (mxschmitt) [inline stores/redis/redis.go]: I would simplify the Variable declarations to `:=` and remove the declaration 2 lines above.

File: stores/redis/redis.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+		return true, nil
+	}
+	logrus.Debugf("Key '%s' does not exist!", key)
+	return false, nil
+}
+
+func (r *RedisStore) setValue(key string, raw []byte) error {
+	var status *redis.StatusCmd
+	logrus.Debugf("Setting value for key '%s: '%s''", key, raw)
+	status = r.c.Set(key, raw, 0) // n.b. expiration 0 means never expire
```
Comment (mxschmitt) [inline stores/redis/redis.go]: I would simplify the Variable declarations to `:=` and remove the declaration 2 lines above.

File: stores/redis/redis.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+		msg := fmt.Sprintf("Could not check existence of key '%s': %s", key, err)
+		logrus.Error(msg)
+		return errors.Wrap(err, msg)
+	}
+	if exists == false {
+		logrus.Warnf("Tried to delete key '%s' but it's already gone", key)
+		return err
+	}
+
+	status = r.c.Del(key)
```
Comment (mxschmitt) [inline stores/redis/redis.go]: I would simplify the Variable declarations to `:=` and remove the declaration 2 lines above.

File: stores/redis/redis.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+	var entry *shared.Entry
+	var visitor *shared.Visitor
+	var lastVisit time.Time
+	var raw []byte
+	var err error
+
+	entryKey := entryPathPrefix + id
+	logrus.Debugf("Fetching key: '%s'", entryKey)
+	result := r.c.Get(entryKey)
+	raw, err = result.Bytes()
```
Comment (mxschmitt) [inline stores/redis/redis.go]: I would simplify the Variable declarations to `:=` and remove the declaration 5 lines above.

File: stores/redis/redis.go [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+	visitCount, err := r.c.LLen(entryVisitsKey).Result()
+	if err != nil {
+		logrus.Warnf("Could not get length of visitor list for id '%s': '%v'", id, err)
+		entry.Public.VisitCount = int(0) // or zero if nobody's visited, that's fine.
+	} else {
+		entry.Public.VisitCount = int(visitCount)
+	}
+
+	// grab the timestamp out of the last visitor on the list
+	lastVisit = time.Time(time.Unix(0, 0)) // default to start-of-epoch if we can't figure it out
```
Comment (mxschmitt) [inline stores/redis/redis.go]: I would simplify the Variable declarations to `:=` and remove the declaration in line 203

**Comment by mxschmitt** [comment]
And thanks for you PR 👍

File: stores/redis/redis.go (hunk 2 above)
Comment (memory) [reply stores/redis/redis.go]: done

File: stores/redis/redis.go (hunk 3 above)
Comment (memory) [reply stores/redis/redis.go]: done

File: stores/redis/redis.go (hunk 4 above)
Comment (memory) [reply stores/redis/redis.go]: done

File: stores/redis/redis.go (hunk 5 above)
Comment (memory) [reply stores/redis/redis.go]: done

File: stores/redis/redis.go (hunk 6 above)
Comment (memory) [reply stores/redis/redis.go]: done

**Comment by memory** [comment]
Okay, the second commit in this PR should address the style concerns.  Also I updated the README and the example config.yaml.

**Comment by mxschmitt** [comment]
As a final step please fix your comment styling to "MethodName ..." https://app.codacy.com/app/maxibanki/golang-url-shortener/pullRequest?prid=1621352

**Comment by memory** [comment]
55311f5 updates the comment styling to sentence format as requested.

**Pull request merged**